geps/gep-1494/index.md

```diff
@@ -130,20 +130,383 @@ From @ongy, some additional goals to keep in mind:
+	// the body of the client request; and if so, how big that body is allowed
+	// to be.
+	//
+	// It is expected that implementations will buffer the request body up to
+	// `forwardBody.maxSize` bytes. Bodies over that size must be rejected with a
+	// 4xx series error (413 or 403 are common examples), and fail processing
+	// of the filter.
+	//
+	// If unset, or `forwardBody.maxSize` is set to `0`, then the body will not
+	// be forwarded.
+	//
+	// Feature Name: HTTPRouteExtAuthForwardBody
+	//
+	// GEP Review Notes:
+	// Both Envoy and Traefik show support for this feature, but HAProxy and
+	// ingress-nginx do not. So this has a separate feature flag for it.
+	//
+	// +optional
+	ForwardBody *ForwardBodyConfig `json:"forwardBody,omitempty"`
+}
+
+// GRPCAuthConfig contains configuration for communication with ext_auth
+// protocol-speaking backends.
+type GRPCAuthConfig struct {
+
+	// AllowedRequestHeaders specifies what headers from the client request
+	// will be sent to the authorization server.
+	//
+	// If this list is empty, then all headers must be sent.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxLength=64
+	AllowedRequestHeaders []string `json:"allowedHeaders,omitempty"`
+}
+
+// HTTPAuthConfig contains configuration for communication with HTTP-speaking
+// backends.
+type HTTPAuthConfig struct {
+	// Path sets the prefix that paths from the client request will have added
+	// when forwarrded to the authorization server.
+	//
+	// When empty or unspecified, no prefix is added.
+	// +optional
+	Path string `json:"path,omitempty"`
+
+	// AllowedRequestHeaders specifies what additional headers from the client request
+	// will be sent to the authorization server.
+	//
+	// The following headers must always be sent to the authorization server,
+	// regardless of this setting:
+	// 
+	// * `Host`
+	// * `Method`
+	// * `Path`
+	// * `Content-Length`
+	// * `Authorization`
+	//
+	// If this list is empty, then only those headers must be sent.
+	//
+	// Nick's GEP Review Notes:
+	// Envoy actually does not allow sending all headers to HTTP authorization
+	// servers. If the equivalent Envoy field is unspecified, then only
+	// `Host`, `Method`, `Path`, `Content-Length` and `Authorization` are included.
+	// This is kept for backwards compatibility in Envoy.
+	//
+	// Traefik includes _all_ headers when the equivalent setting is empty.
+	//
+	// This definition is the only compromise I can see between the two; Envoy
+	// can set this to empty, and Traefik can always explicitly specify those headers
+	// at a minimum.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxLength=64
+	AllowedRequestHeaders []string `json:"allowedHeaders,omitempty"`
+
+	// AllowedResponseHeaders specifies what headers from the authorization response
+	// will be copied into the request to the backend.
+	//
+	// If this list is empty, then all headers from the authorization server
+	// except Authority or Host must be copied.
```

Comment:
I'd vote for being more restrictive than "all headers". We used to default to just

- `Authorization`
- `Location`
- `Proxy-Authenticate`
- `Set-Cookie`
- `WWW-Authenticate`

which is probably a little _too_ restrictive these days... 🤔

Reply:
This is the default for the filter itself, which is why I started from here - if the corresponding field (`allowed_headers`) in the filter is empty for gRPC, then all headers are sent. For HTTP, it defaults to the fields I listed when the field is empty. ( See `allowed_headers` in https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto#extensions-filters-http-ext-authz-v3-extauthz).

There's no reason why we have to do that as well, you're correct. We could say, "when this field is empty, send these fields only", and everyone would be able to do that (by specifying only those headers).

Reply:
Happy to change the empty-string behavior to mandate a particular set of headers instead, does anyone else agree with @kflynn's list here? (I haven't done this sort of config for a _long_ time, so happy to be told what's expected here.).

If I don't hear from anyone in the next day or so, I'll go ahead with that list above.

Reply:
Changed, gRPC now mandates that the headers list being empty means "send the above list instead".